rfcs/text/0013_background_sessions.md

```diff
@@ -0,0 +1,484 @@
+ 2. Kibana cancels queries upon navigating away from an applicaiton, once again, as a mean of protecting clusters and reducing unnecessary load. 
+
+In 7.7, with the introduction of the `_async_search` API in Elasticsearch, we provided Kibana users a way to bypass the timeout, but users still need to remain on-screen for the entire duration of the search requests.
+
+The primary motivation of this RFC is to enable users to do the following without needing to keep Kibana open, or while moving onto other work inside Kibana:
+
+- Run long search requests (beyond 30 seconds)
+- View their status (complete/incomplete)
+- Cancel incomplete search requests
+- Retrieve completed search request results 
+
+# Detailed design
+
+Because a single view (such as a dashboard with multiple visualizations) can initiate multiple search requests, we need a way to associate the search requests together in a single entity. 
+
+We call this entity a `session`, and when a user decides that they want to continue running the search requests while moving onto other work, we will create a saved object corresponding with that specific `session`, persisting the *sessionId* along with a mapping of each *request's hash* to the *async ID* returned by Elasticsearch.
+
+## High Level Flow Charts
+
+### Client side search
+
+This diagram matches any case where `data.search` is called from the front end:
+
+![image](../images/background_sessions_client.png)
+
+### Server side search
+
+This case happens if the server is the one to invoke the `data.search` endpoint, for example with TSVB.
+
+![image](../images/background_sessions_server.png)
+
+## Data and Saved Objects
+
+### Background Session Status
+
+```ts
+export enum BackgroundSessionStatus {
+  Running,   // The session has at least one running search ID associated with it.
+  Done,      // All search IDs associated with this session have completed.
+  Error,     // At least one search ID associated with this session had an error. 
+  Expired,   // The session has expired. Associated search ID data was cleared from ES.
+}
+```
+
+### Saved Object Structure
+
+```ts
+interface BackgroundSessionAttributes extends SavedObjectAttributes {
+  sessionId: string;
+  status: BackgroundSessionStatus;
+  name: string;
+  creation: string;
+  expiration: string;
+  idMapping: { [key: string]: string };
+  url: string; // TODO
+  metadata: { [key: string]: any } // TODO
+}
+```
+
+## Frontend Services
+
+Most sessions will probably not be saved. Therefore, to avoid creating unnecessary saved objects, the browser will keep track of requests and their respective search IDs, until the user chooses to store the session. Once a session is stored, any additional searches will be immediately saved on the server side. 
+
+### New Session Service
+
+We will expose a new frontend `session` service on the `data` plugin `search` service. 
+
+The service will expose the following APIs:
+
+```ts
+interface ISessionService {
```

Comment:
Hmm, totally agreed. @lizozom We might want to sync about this. It might make sense to have any method that is on the service itself require a `sessionId`, and then to have an actual class that represents the background session, and have the other methods (that don't require the `sessionId` on the object itself.

Reply:
wasn't the idea to limit the scope to a single active session ?